lib/loader/file-loader.ts

```diff
@@ -147,17 +134,132 @@
       `File-loader has loaded a configuration file from ${result.filepath}`,
     );
 
-    let config = result.config;
+    return result.config;
+  };
+};
+
+function resolveReference(
+  reference: string,
+  currentContext: Record<string, any>,
+): Record<string, any> | string | number | null {
+  const parts = reference.split('.');
+  let value = currentContext;
 
-    if (!(options.ignoreEnvironmentVariableSubstitution ?? true)) {
-      const replacedConfig = placeholderResolver(
-        JSON.stringify(result.config),
-        process.env,
-        options.disallowUndefinedEnvironmentVariables ?? true,
-      );
-      config = JSON.parse(replacedConfig);
+  for (const part of parts) {
+    value = value[part];
+    if (value === undefined) {
+      return null;
     }
+  }
 
-    return config;
-  };
-};
+  return value;
+}
+
+function transformFileLoaderResult(
+  obj: Record<string, any> | string | number,
+  context: Record<string, any>,
+  disallowUndefinedEnvironmentVariables: boolean,
+  visited = new Set<Record<string, any> | string | number>(),
+): Record<string, any> | string | number {
+  if (typeof obj === 'string') {
+    const match = obj.match(/\$\{(.+?)\}/g);
+    if (match) {
+      for (const placeholder of match) {
+        const variable = placeholder.slice(2, -1);
+        let resolvedValue = resolveReference(variable, context);
+
+        if (obj === resolvedValue) {
+          throw new Error(
+            `Circular self reference detected: ${obj} -> ${resolvedValue}`,
+          );
+        }
+
+        if (resolvedValue !== null) {
+          if (typeof resolvedValue === 'string') {
+            // resolve reference first
+            if (resolvedValue.match(/\$\{(.+?)\}/)) {
+              try {
+                resolvedValue = transformFileLoaderResult(
+                  resolvedValue,
+                  context,
+                  disallowUndefinedEnvironmentVariables,
+                  visited,
+                );
+              } catch (error) {
+                if (error instanceof RangeError) {
+                  debug(
+                    `Can not resolve a circular reference in ${obj} -> ${resolvedValue} -> ${error.message}`,
```

Comment:
I think it is a false positive because it will log the value and the value is a reference to another key always, otherwise we won't be able to come to this part of the code